File: voss/crashpoint.py
"""Deterministic crash-injection points for fault-injection tests (RFC 5.5).

A real crash can land at any byte offset of program progress.  These seams let
a test driver die (``os._exit`` == power loss) at exactly the moments where the
write-ahead ledger boundary matters:

- voss_cp_flow_request_prewal     flow created in memory, nothing recorded yet
- voss_cp_flow_request            flow_request record appended
- voss_cp_flow_resolution         approval granted, flow_resolution appended
- voss_cp_capability_issued       capability_issued appended
- voss_cp_request_executed        request_executed appended (before effect)
- voss_cp_effect_done             tool effect materialized, result not logged

Only active when ``VOSS_CRASHPOINT`` names the point; a worker can never set
it (the trusted process owns its environment).  No effect otherwise.
"""
from __future__ import annotations

import os

CP_FLOW_REQUEST_PREWAL = "voss_cp_flow_request_prewal"
CP_FLOW_REQUEST = "voss_cp_flow_request"
CP_FLOW_RESOLUTION = "voss_cp_flow_resolution"
CP_CAPABILITY_ISSUED = "voss_cp_capability_issued"
CP_REQUEST_EXECUTED = "voss_cp_request_executed"
CP_EFFECT_DONE = "voss_cp_effect_done"

_EXIT_CODE = 86  # distinctive: means "the intended crashpoint fired"


def maybe_crash(point: str) -> None:
    if os.environ.get("VOSS_CRASHPOINT") == point:
        os._exit(_EXIT_CODE)  # no cleanup, no flush guarantees beyond emit()

File: rust/src/crashpoint.rs
//! Deterministic crash-injection seams for fault-injection tests.
//!
//! A crash can land at any byte of progress. These seams stop the process
//! (`std::process::exit`, with no destructor or flush beyond what the ledger
//! already committed) at the write-ahead boundaries that matter. A seam fires
//! only when `VOSS_CRASHPOINT` names it, or when this process armed that same
//! name. The worker cannot arm the trusted host.

use std::sync::Mutex;

pub const FLOW_REQUEST_PREWAL: &str = "voss_cp_flow_request_prewal";
pub const FLOW_REQUEST: &str = "voss_cp_flow_request";
pub const FLOW_RESOLUTION: &str = "voss_cp_flow_resolution";
pub const CAPABILITY_ISSUED: &str = "voss_cp_capability_issued";
pub const REQUEST_EXECUTED: &str = "voss_cp_request_executed";
pub const EFFECT_DONE: &str = "voss_cp_effect_done";

pub const EXIT_CODE: i32 = 86;

static ARMED: Mutex<Option<String>> = Mutex::new(None);

pub fn arm(point: &str) {
    *ARMED.lock().expect("crashpoint") = Some(point.to_string());
}

pub fn maybe_crash(point: &str) {
    let from_env = std::env::var("VOSS_CRASHPOINT").ok();
    let armed = ARMED.lock().expect("crashpoint").clone();
    if from_env.as_deref() == Some(point) || armed.as_deref() == Some(point) {
        std::process::exit(EXIT_CODE);
    }
}
